File: LeetCode/Medium/3689-maximum-total-subarray-value-i/3689-maximum-total-subarray-value-i.java
class Solution {
    public long maxTotalValue(int[] nums, int k) {
        /*
         * Calculates the maximum possible total value by choosing k subarrays.
         * * Approach:
         * 1. Because the problem statement explicitly mentions subarrays can overlap 
         * and be chosen more than once, we don't need complex dynamic programming.
         * 2. The best strategy is purely greedy: find the single maximum possible 
         * subarray value, and reuse it 'k' times.
         * 3. The maximum subarray value achievable is (global_max - global_min).
         * 4. To avoid integer overflow during multiplication, the difference must 
         * be explicitly cast to a 'long'.
         * * Complexity:
         * - Time Complexity: O(n) - One pass loop through the array.
         * - Space Complexity: O(1) - Constant memory allocation.
         */
        int max = Integer.MIN_VALUE;
        int min = Integer.MAX_VALUE;
        
        // Single pass to find both limits efficiently without stream overhead
        for (int num : nums) {
            if (num > max) max = num;
            if (num < min) min = num;
        }
        
        // Safe casting to prevent overflow
        return (long)(max - min) * k;
    }
}

File: LeetCode/Medium/3689-maximum-total-subarray-value-i/3689-maximum-total-subarray-value-i.py
class Solution:
    def maxTotalValue(self, nums: list[int], k: int) -> int:
        """
        Calculates the maximum possible total value by choosing k subarrays.
        
        KEY INSIGHT:
        The problem states that "the exact same subarray can be chosen more than once." 
        To maximize our final total, we don't need a complex splitting strategy. 
        Instead, we should find the single absolute best subarray that yields the 
        highest possible value, and greedily choose that exact same subarray 'k' times.
        
        The maximum value of any subarray is bounded by the difference between 
        the highest and lowest numbers in the entire array (global_max - global_min). 
        Thus, the entire problem boils down to: (global_max - global_min) * k.
        
        Complexity:
        - Time Complexity: O(n) because we find the max and min in a single pass.
        - Space Complexity: O(1) auxiliary space.
        """
        # Find the global maximum and minimum in O(n) time
        global_max = max(nums)
        global_min = min(nums)
        
        # Replicating the absolute best subarray 'k' times
        return (global_max - global_min) * k